### PR TITLE
[AIR/Data] Allow for nested Chain preprocessors

### LINKED ISSUE #28618: [AIR] Allow for nested Chain preprocessors
### Description

Allow for Chains to be a part of other Chains. This is explicitly disabled right now, but I don't see a reason why it would need to be.

### Use case

This is a common pattern in scikit-learn - have a pipeline (chain) composed of pipelines. Eg. a categorical pipeline for categorical columns and a numeric pipeline for the rest.

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

## Why are these changes needed?

Allows for Chain preprocessors to be nested. This was disabled, but I am not sure what was the reason for it, as everything seems to be working fine. Nesting pipelines is a common pattern in scikit-learn.

## Related issue number

Closes https://github.com/ray-project/ray/issues/28618

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by clarkzinzow — APPROVED** [review]
LGTM, @matthewdeng do you remember why we didn't support nested chained preprocessors initially?

**Comment by matthewdeng** [comment]
@clarkzinzow I don't remember the exact reason at the time, but from my point of view:

* I do see a benefit of being able to group preprocessors in a Chain and then Chaining groups of preprocessors.
* I have some concerns that nested preprocessors can lead to complexity, e.g. introducing circular dependencies.

Generally I would lean towards keeping the design simple with a flattened list of Preprocessors, until there is a compelling need to support this.

**Comment by Yard1** [comment]
@matthewdeng I think that we shouldn't prevent users from nesting preprocessors. By doing that, we are trying to solve a problem that's not there, and we can see in scikit-learn that nesting pipelines is a common occurrence. In fact, I would say that we are increasing complexity by having the check in the first place.

File: python/ray/data/preprocessors/chain.py [hunk 1]
```diff
@@ -45,16 +45,13 @@ def fit_status(self):
...
-            assert not isinstance(
-                p, Chain
-            ), "A chain preprocessor should not contain another chain preprocessor."
             if p.fit_status() == Preprocessor.FitStatus.FITTED:
                 fittable_count += 1
                 fitted_count += 1
-            elif p.fit_status() == Preprocessor.FitStatus.NOT_FITTED:
+            elif p.fit_status() in (
+                Preprocessor.FitStatus.NOT_FITTED,
+                Preprocessor.FitStatus.PARTIALLY_FITTED,
```
Comment (matthewdeng) [inline python/ray/data/preprocessors/chain.py:53]: Can we double check this doesn't lead to any incorrect state.

**Review by matthewdeng — APPROVED** [review]
For posterity, this was originally restricted here: https://github.com/ray-project/ray/pull/23700#discussion_r844233366

**Comment by clarkzinzow** [comment]
@Yard1 Is this good to merge?

**Comment by Yard1** [comment]
@clarkzinzow Yes!

**Pull request merged**